contracts/adapters/UniswapAdapter.sol

```diff
@@ -0,0 +1,209 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.7.2;
+pragma experimental ABIEncoderV2;
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+
+import {IUniswapV2Pair} from "../interfaces/IUniswapV2Pair.sol";
+import {IUniswapV2Router02} from "../interfaces/IUniswapV2Router.sol";
+
+contract UniswapAdapter {
+
+    enum Exchange {Uniswap, Sushiswap}
+
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+
+    address public immutable ethAddress;
+    address public immutable diggAddress;
+    address public constant wethAddress = address(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2);
+    string private constant _name = "UNISWAP";
+    bytes32 private constant uniswapHash = keccak256(abi.encodePacked("UNISWAP"));
+    bytes32 private constant sushiswapHash = keccak256(abi.encodePacked("SUSHISWAP"));
+    address public immutable wbtcAddress;
+    bool private constant _nonFungible = true;
+    IUniswapV2Router02 public immutable uniswapRouter;
+    IUniswapV2Router02 public immutable sushiswapRouter;
+    IUniswapV2Pair public immutable wbtcDiggUniswap;
+    IUniswapV2Pair public immutable wbtcDiggSushiswap;
+    IERC20 public immutable wbtcToken;
+    uint256 public constant deadlineBuffer = 150;
+
+    constructor(
+        address _uniswapRouter,
+        address _sushiswapRouter,
+        address _wbtcAddress,
+        address _ethAddress,
+        address _wbtcDiggUniswap,
+        address _wbtcDiggSushiswap,
+        address _diggAddress
+    ) public {
+
+        wbtcAddress = _wbtcAddress;
+        ethAddress = _ethAddress;
+        diggAddress = _diggAddress;
+        uniswapRouter = IUniswapV2Router02(_uniswapRouter);
+        sushiswapRouter = IUniswapV2Router02(_sushiswapRouter);
+        wbtcDiggUniswap = IUniswapV2Pair(_wbtcDiggUniswap);
+        wbtcDiggSushiswap = IUniswapV2Pair(_wbtcDiggSushiswap);
+        wbtcToken = IERC20(_wbtcAddress);
+
+        SafeERC20.safeApprove(IERC20(_wbtcAddress), address(_uniswapRouter), type(uint256).max);
+        SafeERC20.safeApprove(IERC20(_wbtcAddress), address(_sushiswapRouter), type(uint256).max);
+        SafeERC20.safeApprove(IERC20(_diggAddress), address(_uniswapRouter), type(uint256).max);
+        SafeERC20.safeApprove(IERC20(_diggAddress), address(_sushiswapRouter), type(uint256).max);

```

Comment:
Since we're doing a [delegatecall](https://medium.com/coinmonks/delegatecall-calling-another-contract-function-in-solidity-b579f804178c) for the adapters, the approval will only be for the Adapter's logic contract, but not the actual instrument contract.

We can completely remove this., and approve max on-the-fly.